jOOL-java-8/src/main/java/org/jooq/lambda/Agg.java

```diff
@@ -979,4 +979,694 @@ else if (percentile == 1.0)
+     * @version 1.0
+     * @author Jichen Lu
+     * @date 2021-05-17
+     */
+    public static <T> Collector<T, ?, Optional<Double>> regrR2By(Function<? super T, ? extends Number> functionX, Function<? super T, ? extends Number> functionY) {
+        class Accumulator {
+            Integer n = 0;
+            Double sumXY = 0.0;
+            Double sumXX = 0.0;
+            Double sumYY = 0.0;
+            Double sumX = 0.0;
+            Double sumY = 0.0;
+
+            ArrayList<Double> listX = new ArrayList<>();
+            ArrayList<Double> listY = new ArrayList<>();
+        }
+
+        return Collector.of(
+                () -> new Accumulator(),
+                (a, t) -> {
+                    Double x = (Double) functionX.apply(t);
+                    Double y = (Double) functionY.apply(t);
+                    if (x != null && y != null) {
+                        a.listX.add(x);
+                        a.listY.add(y);
+                        a.n++;
+                        a.sumXY += x * y;
+                        a.sumXX += x * x;
+                        a.sumYY += y * y;
+                        a.sumX += x;
+                        a.sumY += y;
+                    }
+                },
+                (a1, a2) -> {
+                    a1.n += a2.n;
+                    a1.listX.addAll(a2.listX);
+                    a1.listY.addAll(a2.listY);
+                    a1.sumX += a2.sumX;
+                    a1.sumY += a2.sumY;
+                    a1.sumXX += a2.sumXX;
+                    a1.sumYY += a2.sumYY;
+                    a1.sumXY += a2.sumXY;
+                    return a1;
+                },
+                a -> {
+
+                    if (a.n == 0 || a.n == 1) {
+                        return Optional.empty();
+                    }
+
+
+                    Double slope = 0.0;
+                    Double intercept = 0.0;
+                    if ((a.sumXX - a.sumX * a.sumX / a.n) == 0.0) {
+                        intercept = a.sumY / a.n;
+                        slope = 0.0;
+                    } else {
+                        slope = ((a.sumXY - a.sumX * a.sumY / a.n) / (a.sumXX - a.sumX * a.sumX / a.n));
+                        intercept = a.sumY / a.n - (a.sumXY - a.sumX * a.sumY / a.n) / (a.sumXX - a.sumX * a.sumX / a.n) * a.sumX / a.n;
+                    }
+
+                    Double avgY = a.sumY / a.n;
+                    Double sstot = 0.0;
+                    Double ssreg = 0.0;
+                    for (int i = 0; i < a.listY.size(); i++) {
+                        sstot += Math.pow((a.listY.get(i) - avgY), 2);
+                        ssreg += Math.pow(slope * a.listX.get(i) + intercept - a.listY.get(i), 2);
+                    }
+                    if (sstot == 0.0) {
+                        return Optional.of(1.0);
+                    } else {
+                        return Optional.of(1 - ssreg / sstot);
+                    }
+
+                }
+        );
+    }
```

Comment:
I'm not going to review all the formulas as implemented here, but I'm positive that a lot more code reuse is possible. Linear regression functions tend to depend on one another, or on other statistical functions, such as `COUNT()`, `AVG()`, `VAR_POP()`, `COVAR_POP()`, `STDDEV_POP()`, etc. etc. (see formulas here: https://github.com/jOOQ/jOOQ/issues/11547).

I'd much rather there were many more composable building blocks improving:

- Code reuse
- Allowing for things to be calculated for any data type, with only a single implementation.

Reply:
This is still my main concern here. There's *a lot* of repeated maths, which I find hard to review and maintain. Maybe it can be refactored?